### PR TITLE
Add a new domain and implementation for ProgramWorkflow

### PR DESCRIPTION
**Pull request by ridmal**
- Add programWorkflow CSV parsers
- Add tests for programWorkflow loading
- Add CSV file to keep the programWorkflow meta-data

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/initializer/InitializerConstants.java [hunk 1]
```diff
@@ -51,4 +51,6 @@
 	
 	public static final String DOMAIN_PROG = "programs";
 	
+	public static final String DOMAIN_PROG_WF = "programWorkflows";
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/InitializerConstants.java]: The convention for the values here is 'no case' (look at the others), make it `"programworkflows"`:
```java
public static final String DOMAIN_PROG_FL = "programsworkflows";
```

File: api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowCsvParser.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	protected boolean isVoidedOrRetired(ProgramWorkflow instance) {
+		return instance.isRetired();
+	}
+	
+	@Override
+	protected ProgramWorkflow doVoidRetireActions(ProgramWorkflow instance) {
+		ProgramWorkflow programWorkflow = service.getWorkflowByUuid(instance.getUuid());
+		if (programWorkflow != null) {
+			Program program = instance.getProgram();
+			program.retireWorkflow(instance);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowCsvParser.java]: At this point `instance` is already marked as retired and will be saved as such later down the line anyway. Are you not supposed to simply do:
```java
program.removeWorkflow(instance);
```
?

File: api/src/main/java/org/openmrs/module/initializer/api/CsvParser.java [hunk 3]
```diff
@@ -160,6 +160,13 @@ protected void close() throws IOException {
 		reader.close();
 	}
 	
+	/*
+	 * Action for void/retire instances.
+	 */
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/CsvParser.java]: Make it a JavaDoc, and more explicit:
```java
/**
 * This method provides a space to perform actions after setting an object as voided or retired.
 *
 * Eg.: A program workflow needs to be removed from its program after being marked as retired. 
 * @see ProgramWorkflowCsvParser#doVoidRetireActions(org.openmrs.ProgramWorkflow)
 */
```

File: api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowLineProcessor.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import org.openmrs.api.context.Context;
+import org.openmrs.module.initializer.api.BaseLineProcessor;
+import org.openmrs.module.initializer.api.CsvLine;
+import org.openmrs.module.initializer.api.impl.Utils;
+
+public class ProgramWorkflowLineProcessor extends BaseLineProcessor<ProgramWorkflow, ProgramWorkflowService> {
+	
+	protected static String HEADER_PROGRAM = "program";
+	
+	protected static String HEADER_CONCEPT = "concept";
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowLineProcessor.java]:
```java
protected static String HEADER_WORKLOW_CONCEPT = "workflow concept";
```

File: api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+import org.openmrs.test.BaseModuleContextSensitiveTest;
+import org.openmrs.test.Verifies;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
+
+public class DomainProgramWorkFlowsInitializerServiceTest extends DomainBaseModuleContextSensitiveTest {
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java]: Just thinking, wouldn't you have a lot already available to help you run tests if you were to inherit `DomainProgramsInitializerServiceTest` and rerun its setup in this test's setup?
I'm not sure, you need to try. At least you could expose some test utils that one can reuse from the other.

File: api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+			Assert.assertEquals(cs.getConceptByName("concept1"), programWorkflow.getConcept());
+			Assert.assertEquals(false, programWorkflow.isRetired());
+			
+			Program program = pws.getProgramByName("program1");
+			Assert.assertEquals(programWorkflow, program.getWorkflow(programWorkflow.getId()));
+			
+		}
+		{
+			ProgramWorkflow programWorkflow = pws.getWorkflowByUuid("2b98bc76-245c-11e1-9cf0-00248140a5ef");
+			Assert.assertNull(programWorkflow);
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java]: What is this test showing exactly?

File: api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		// A Program used for add ProgramWorkflows
+		{
+			Program pro = new Program();
+			pro.setConcept(cs.getConceptByName("programConcept"));
+			pro.setOutcomesConcept(cs.getConceptByName("outcomesConceptTest1"));
+			pro.setName("program1");
+			pro.setDescription("Program Description");
+			pro = pws.saveProgram(pro);
+		}
+		// A programWorkFlow to be edited
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java]: Do you mean?
>`// A workflow to be removed from its program`

File: api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java (hunk 6 above)
Comment (ridmal) [reply api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java]: This is for showing voided programWorkflows are not adding to the database.

File: api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowCsvParser.java (hunk 2 above)
Comment (ridmal) [reply api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowCsvParser.java]: Actually first i used that. But in openMRS core , when we remove a programWorkflow from a program, it just changed as retired. But like you said it already done in the ```programWorkflow.setRetired(getVoidOrRetire(line.asLine()));``` we can skip that.

File: api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java (hunk 6 above)
Comment (mks-d) [reply api/src/test/java/org/openmrs/module/initializer/api/DomainProgramWorkFlowsInitializerServiceTest.java]: Ok, we never were adding such tests until now, but then leave it ;-)

However we also need a real case of _edition_ to be tested, showing that changing the defining concept is well taken into account.

File: api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowCsvParser.java (hunk 2 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/module/initializer/api/programWorkflow/ProgramWorkflowCsvParser.java]: Can you detail exactly what each of
- `program.retireWorkflow(instance)`
- `program.removeWorkflow(instance)`

does?

**Pull request merged**